### DISCUSSION TITLE
Proposal: `using` directives everywhere

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#218] [General]
Originally proposed at dotnet/roslyn#10299, dotnet/roslyn#4491

The idea is to be able to use `using` directives in any block,
```cs
class X
{
  using Self = X; // type level (1)
  void F()
  {
    using Foo.Extensions; // block level (2)
    using static System.Console;
  }
}
```
This would enable (1) type-scoped type aliases and (2) not importing types to the whole file when they are just required in a particular block. 

Closer usings should take precedence in binding as they already do for global and namespace-level usings.

**Open question:** should we also support `enum` blocks in addition to class, struct and interface?
```cs
enum E {
  using static Other;
  A = X, // Other.X;
}
```
Since we do have a championed proposal for enum members (https://github.com/dotnet/csharplang/issues/297), I think it would make sense to have it.

**Comment by HaloFour** [comment]
I might suggest a change to the name of this proposal/discussion to make it clear that it's referring to the `using` directive for namespace resolution and aliasing vs. the `using` statement for disposable resources or unit-of-work.

**Comment by MillKaDe** [comment]
1) It would be nice, if a type-local using alias could also be used outside of its defining type like this:
```c#
using SCG = System.Collections.Generic; // file-global scope
namespace X {
  class Alpha {
    using AQS = SCG.Stack<SCG.Queue<Action>>;
    // ...
  } // class Alpha
  class Beta {
    Alpha.AQS actions; // uses alias 'AQS' defined above in class Alpha
  } // class Beta
} // namespace X
```

2) In addtion to more local scoping of using alias, more global scoping would sometimes be useful too.
Maybe a using alias in one file of a project could 'exported' to the whole project or solution by marking it with an attribute ?
```c#
[AliasExportToProject]
using SCG = System.Collections.Generic; // file-global scope
```
This would allow to use SCG in all files of the same project without redefining it in every file.

```c#
[AliasExportToSolution]
using SCG = System.Collections.Generic; // file-global scope
```
This would allow to use SCG in all files of all projects of the same solution without redefining it in every file.

```c#
namespace XYZ {
  [AliasExportToNamespaceUse]
  using SCG = System.Collections.Generic; // namespace scope
} // namespace XYZ
```
This would allow to use SCG in all files of all projects of all solutions, but only in those scopes where `using XYZ` has been defined.

I think these additional 'exports' could cover all cases, where C# using alias is more limited than C/C++ `typedef`.

**Comment by alrz** [comment]
@MillKaDe 

I think you should open another issue for that feature since it's divergent to the subject of this thread.

**Comment by JasonBock** [comment]
I kind of like this. I can see where this would be beneficial where you have type name collisions and you want to be specific with a type in a block of code. I can also see where this would be confusing. Interesting idea.

**Comment by lachbaer** [comment]
I wouldn't have it on namespace level, because it implies that the imported types and namespaces are available to the same namespace in another file, too, even if the using directive is not repeated there.

**Comment by HaloFour** [comment]
@lachbaer

That's been permitted since C# 1.0.

**Comment by lachbaer** [comment]
@HaloFour Okay, never used it and never seen it actually...

**Comment by alrz** [comment]
All built-in templates in VS use global using directives. plus roslyn. What IDE do you use?

**Comment by lachbaer** [comment]
Global using, before the namespace declaration, but not after it.

**Comment by alrz** [comment]
>it implies that the imported types and namespaces are available to the same namespace in another file, too, even if the using directive is not repeated there.

Just to be clear, that's why it's called using "directive" and not declaration. It only affects the current file.

**Comment by lachbaer** [comment]
But isn't the intention of this proposal to *scope* that directive to its block? I'd find it confusing when being used within namespace *scope*. That's probably why I haven't seen it being used by others yet.

If this feature were available I would make use of it primarily on subclasses and methods.

**Comment by alrz** [comment]
Namespace-level using directives already have a scope: the enclosing namespace.
```cs
namespace A { using System; }
namespace B { /* System is not imported here. */ }
```
The intention of this proposal is to extend that to other blocks, in addition to namespace's.

**Comment by lachbaer** [comment]
#133 should respect this proposal as well.

**Comment by lachbaer** [comment]
> Namespace-level using directives already have a scope: the enclosing namespace.

Now I know :smile: Nevertheless I find that confusing and won't use it. But this proposal I really like.

**Comment by alrz** [comment]
To not conflict with implicitly scoped usings (https://github.com/dotnet/csharplang/issues/1174), we probably want something to disambiguate. like
```cs
void M() {
  using namespace My.Extensions;
  using static System.Console;
  using alias MyType = Some.Ambiguous.Type;

}
```
Since this is a statement-level directive, that way it stands out as a directive among statements.

**Comment by tilkinsc** [comment]
#7255
